libdd-profiling-ffi/src/profiles/profiles_dictionary.rs

```diff
@@ -0,0 +1,261 @@
+// Copyright 2025-Present Datadog, Inc. https://www.datadoghq.com/
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::arc_handle::ArcHandle;
+use crate::profile_status::ProfileStatus;
+use crate::profiles::utf8::Utf8Option;
+use crate::profiles::{
+    ensure_non_null_insert, ensure_non_null_out_parameter, null_profiles_dictionary,
+};
+use crate::ProfileError;
+use libdd_common_ffi::slice::CharSlice;
+use libdd_profiling::profiles::collections::StringRef;
+use libdd_profiling::profiles::datatypes::{
+    Function2, FunctionId2, Mapping2, MappingId2, ProfilesDictionary, StringId2,
+};
+
+/// A StringId that represents the empty string.
+/// This is always available in every string set and can be used without
+/// needing to insert it into a string set.
+#[no_mangle]
+pub static DDOG_PROF_STRINGID2_EMPTY: StringId2 = StringId2::EMPTY;
+
+/// A StringId that represents the string "end_timestamp_ns".
+/// This is always available in every string set and can be used without
+/// needing to insert it into a string set.
+#[no_mangle]
+pub static DDOG_PROF_STRINGID2_END_TIMESTAMP_NS: StringId2 =
+    unsafe { core::mem::transmute(StringRef::END_TIMESTAMP_NS) };
+
+/// A StringId that represents the string "local root span id".
+/// This is always available in every string set and can be used without
+/// needing to insert it into a string set.
+#[no_mangle]
+pub static DDOG_PROF_STRINGID2_LOCAL_ROOT_SPAN_ID: StringId2 =
+    unsafe { core::mem::transmute(StringRef::LOCAL_ROOT_SPAN_ID) };
+
+/// A StringId that represents the string "trace endpoint".
+/// This is always available in every string set and can be used without
+/// needing to insert it into a string set.
+#[no_mangle]
+pub static DDOG_PROF_STRINGID2_TRACE_ENDPOINT: StringId2 =
+    unsafe { core::mem::transmute(StringRef::TRACE_ENDPOINT) };
+
+/// A StringId that represents the string "span id".
+/// This is always available in every string set and can be used without
+/// needing to insert it into a string set.
+#[no_mangle]
+pub static DDOG_PROF_STRINGID2_SPAN_ID: StringId2 =
+    unsafe { core::mem::transmute(StringRef::SPAN_ID) };
+
+/// Allocates a new `ProfilesDictionary` and writes a handle to it in `handle`.
+///
+/// # Safety
+///
+/// - `handle` must be non-null and valid for writes of `ProfilesDictionaryHandle`.
+/// - The returned handle must eventually drop the resource; see
+///   [`ddog_prof_ProfilesDictionary_drop`] for more details.
+/// - If you need a copy, use [`ddog_prof_ProfilesDictionary_try_clone`]; don't just memcpy a new
+///   handle.
+#[no_mangle]
+pub unsafe extern "C" fn ddog_prof_ProfilesDictionary_new(
+    handle: *mut ArcHandle<ProfilesDictionary>,
```

Comment:
This is the new pattern doug wanted?

Reply:
Yes. Consistent return type where out parameters are used instead of union types. This reduces the number of bespoke types since C doesn't have generics, and reduces the boilerplate. It also avoids unions, which are uniquely complex in Java and .NET.